1_data_prep/14_create_train_val_split.py

```diff
@@ -0,0 +1,466 @@
+
+    Args:
+        trees_gdf: GeoDataFrame of individual trees
+        plot_ids: List of plot IDs
+        plot_weights: Dictionary mapping plot_id to weight (number of pairs)
+        use_lumped: If True, use 'lumped_species' column, else use 'species_code'
+    """
+    # Filter trees to only include plots we're working with
+    trees_filtered = trees_gdf[trees_gdf["plot_id"].isin(plot_ids)].copy()
+
+    species_col = "lumped_species" if use_lumped else "species_code"
+
+    print(f"Creating species matrix using: {species_col}")
+    print(f"Found {len(trees_filtered)} individual trees across {len(plot_ids)} plots")
+
+    # Count trees by plot and species
+    species_counts = (
+        trees_filtered.groupby(["plot_id", species_col])
+        .size()
+        .reset_index(name="tree_count")
+    )
+
+    # Pivot to create matrix: rows = plots, columns = species, values = tree counts
+    species_matrix = species_counts.pivot(
+        index="plot_id", columns=species_col, values="tree_count"
+    ).fillna(0)
+
+    # Reindex to match plot order and convert to integer
+    species_matrix = species_matrix.reindex(plot_ids).fillna(0).astype(int)
+
+    # Apply weights - multiply each plot's counts by its weight
+    for plot_id in species_matrix.index:
+        weight = plot_weights.get(plot_id, 1)
+        species_matrix.loc[plot_id] = species_matrix.loc[plot_id] * weight
+
+    print(
+        f"Species quantity matrix: {len(species_matrix)} plots x {len(species_matrix.columns)} species"
+    )
+    print(f"Matrix weighted by number of drone pairs per plot")
+
+    return species_matrix
+
+
+def find_best_split_with_trees(
+    plots_gdf, species_matrix, plot_weights, val_frac=VAL_FRACTION, seed=SEED
+):
+    """
+    Find the best train/val split that minimizes species composition differences.
+    Uses weighted contributions based on number of drone pairs per plot.
+    """
+    np.random.seed(seed)
+
+    plot_ids = plots_gdf["plot_id"].tolist()
+
+    # Calculate target number of datasets for validation (not plots)
+    total_pairs = sum(plot_weights.values())
+    target_val_pairs = int(total_pairs * val_frac)
+
+    best_split = None
+    best_score = float("inf")
+
+    # Try a large number of random splits to find the best one
+    print(f"\nFinding optimal split:")
+    print(f"Total dataset pairs: {total_pairs}")
+    print(f"Target validation pairs: {target_val_pairs} ({val_frac:.0%})")
+    print(f"Testing 100000 different train/val combinations...")
+
+    for i in range(100000):
+        if i % 200 == 0 and i > 0:
+            print(f"Tested {i} combinations (best score so far: {best_score:.4f})")
+
+        # Randomly shuffle and select plots until we reach target validation pairs
+        shuffled_plots = np.random.permutation(plot_ids)
+        val_plots = []
+        val_pairs_count = 0
+
+        for plot_id in shuffled_plots:
```

Comment:
If you wanted a non-iterative version you could use something like this [function ](https://numpy.org/devdocs/reference/generated/numpy.cumsum.html) to get the cumulative sum at each index. I suspect it might be a little faster and cleaner, but I'm not sure.